src/sentry/users/api/serializers/authenticator.py

```diff
@@ -0,0 +1,100 @@
+    authId: str
+    createdAt: datetime
+    lastUsedAt: datetime | None
+
+
+class AuthenticatorInterfaceSerializerResponse(EnrolledAuthenticatorInterfaceSerializerResponse):
+    id: str
+    name: str
+    description: str
+    rotationWarning: str
+    enrollButton: str
+    configureButton: str
+    removeButton: str
+    isBackupInterface: bool
+    isEnrolled: bool
+    disallowNewEnrollment: bool
+    status: str
+    canValidateOtp: bool
+    allowMultiEnrollment: bool
+    allowRotationInPlace: bool
+
+
+@register(AuthenticatorInterface)
+class AuthenticatorInterfaceSerializer(Serializer):
+    def serialize(
+        self,
+        obj: AuthenticatorInterface,
+        attrs: Mapping[str, Any],
+        user: User,
+        **kwargs: Any,
+    ) -> AuthenticatorInterfaceSerializerResponse:
+        data: AuthenticatorInterfaceSerializerResponse = {
+            "id": str(obj.interface_id),
+            "name": str(obj.name),
+            "description": str(obj.description),
+            "rotationWarning": str(obj.rotation_warning),
+            "enrollButton": str(obj.enroll_button),
+            "configureButton": str(obj.configure_button),
+            "removeButton": str(obj.remove_button),
+            "isBackupInterface": obj.is_backup_interface,
+            "isEnrolled": obj.is_enrolled(),
+            "disallowNewEnrollment": obj.disallow_new_enrollment,
+            "status": str(obj.status.value),
+            "canValidateOtp": obj.can_validate_otp,
+            "allowMultiEnrollment": obj.allow_multi_enrollment,
+            "allowRotationInPlace": obj.allow_rotation_in_place,
+        }
+
+        # authenticator is enrolled
+        if obj.authenticator is not None:
+            data["authId"] = str(obj.authenticator.id)
+            data["createdAt"] = obj.authenticator.created_at
+            data["lastUsedAt"] = obj.authenticator.last_used_at
+
+        return data
+
+
+class SmsInterfaceSerializerResponse(AuthenticatorInterfaceSerializerResponse):
+    phone: str
+
+
+@register(SmsInterface)
+class SmsInterfaceSerializer(AuthenticatorInterfaceSerializer):
+    def serialize(
+        self,
+        obj: AuthenticatorInterface,
+        attrs: Mapping[str, Any],
+        user: User,
+        **kwargs: Any,
+    ) -> SmsInterfaceSerializerResponse:
+        data = cast(SmsInterfaceSerializerResponse, super().serialize(obj, attrs, user))
+        assert isinstance(
+            obj, SmsInterface
+        ), "Interface must be SmsInterface to serialize phone number"
+        data["phone"] = obj.phone_number
+        return data
+
+
+for interface in RecoveryCodeInterface, TotpInterface, U2fInterface:
+    register(interface)(AuthenticatorInterfaceSerializer)
+
+
+def get_interface_serializer(interface: AuthenticatorInterface) -> AuthenticatorInterfaceSerializer:
+    if isinstance(interface, SmsInterface):
+        return SmsInterfaceSerializer()
+    return AuthenticatorInterfaceSerializer()
```

Comment:
In the endpoints that are given an `ApplicationInterface`, interface, to serialize, this 'getter' is used to make sure we're using the right serializer. It's a weird scenario because registering the serializer in line 76 and 94 should solve this for us, but because we also have that bug of the file not being imported and therefore the serializers not being registered, this redundant function is here.....

Maybe other alternatives could be having barrel file for just the authenticator ?

Reply:
I don't hate the factory function. In the longer term, I think we should consider removing the `@register()` calls as they contribute to import tangle and import side-effects create hidden dependencies between modules.

> It's a weird scenario because registering the serializer in line 76 and 94 should solve this for us, but because we also have that bug of the file not being imported and therefore the serializers not being registered, this redundant function is here.....

If the endpoint imports the `sentry.users.api.serializers.authenticators` module I would expect that the `register()` calls would behave correctly. Is that not the case?